mne/transforms.py

```diff
@@ -1521,3 +1523,225 @@ def _euler_to_quat(euler):
+    if isinstance(pipeline, str):  # use defaults
+        pipeline = pipeline_defaults[pipeline]
+    for ii, step in enumerate(pipeline):
+        name = f'pipeline[{ii}]'
+        _validate_type(step, str, name)
+        _check_option(name, step, _ORDERED_STEPS)
+    ordered_pipeline = tuple(sorted(
+        pipeline, key=lambda x: _ORDERED_STEPS.index(x)))
+    if pipeline != ordered_pipeline:
+        raise ValueError(
+            f'Steps in pipeline are out of order, expected {ordered_pipeline} '
+            f'but got {pipeline} instead')
+    if len(set(pipeline)) != len(pipeline):
+        raise ValueError('Steps in pipeline should not be repeated')
+    if pipeline[-1] == 'sdr':
+        return tuple(pipeline)[:-1], True
+    return tuple(pipeline), False
+
+
+def _affine_registraion(moving, static, pipeline, niter, zooms):
+    with np.testing.suppress_warnings():
+        from dipy.align import (affine_registration, center_of_mass,
+                                translation, rigid, affine, resample)
+    # optimize transform
+    sigmas = [3.0, 1.0, 0.0]
+    factors = [4, 2, 1]
+    pipeline_options = dict(translation=[center_of_mass, translation],
+                            rigid=[rigid], affine=[affine])
+    for i, step in enumerate(pipeline):
+        # reslice image with zooms
+        logger.info(f'Reslicing to zooms={zooms[step]}...')
+        if i == 0 or zooms[step] != zooms[pipeline[i - 1]]:
+            static_zoomed, static_affine = _reslice_normalize(
+                static, zooms[step])
+        # must be resliced after being moved during every step
+        moving_zoomed, moving_affine = _reslice_normalize(
+            moving, zooms[step])
+        logger.info(f'Optimizing {step}:')
+        with wrapped_stdout(indent='    ', cull_newlines=True):
+            reg_affine = affine_registration(
+                moving_zoomed, static_zoomed, moving_affine, static_affine,
+                nbins=32, metric='MI', pipeline=pipeline_options[step],
+                level_iters=niter[step], sigmas=sigmas, factors=factors)[1]
+
+        # apply the current affine to the full-resolution data
+        moving = resample(_get_img_fdata(moving),
+                          _get_img_fdata(static),
+                          moving.affine, static.affine,
+                          reg_affine)
+
+        # report some useful information
+        if step in ('center_of_mass', 'translation', 'rigid'):
+            dist = np.linalg.norm(reg_affine[:3, 3])
+            angle = np.rad2deg(_angle_between_quats(
+                np.zeros(3), rot_to_quat(reg_affine[:3, :3])))
+            logger.info(f'    Translation: {dist:6.1f} mm')
+            if step == 'rigid':
+                logger.info(f'    Rotation:    {angle:6.1f}°')
+        r2 = _compute_r2(_get_img_fdata(static), _get_img_fdata(moving))
+        logger.info(f'    R²:          {r2:6.1f}%')
+    return reg_affine
```

Comment:
This is going to be incorrect I think, it needs to combine all affines computed at each step. (This arises because you now resample at each step, rather than always working from the starting moving image.)

Reply:
The way it resamples combines the affines ASAIK

Reply:
To my knowledge this is because each step is inclusive of all the previous steps so when you do the rigid registration on the image that has already been translated, the translation affine is included

Reply:
If you take a good T1 image, translate it by 10 mm (you can just set its `affine` to be something different), then apply this function is the output affine a translation by 10 mm? That is what you should get be for both `pipeline=('translate',)` and `pipeline=('translate', 'rigid')`, but I suspect that with the current code if it won't be (because you apply the translation in the first step, so no adjustment is necessary in the next). I also don't 100% know it would have worked on my commit, but I suspect it would have. I can check tomorrow unless you still have time to check today. It would be good to manually check the R**2 values of the actually transformed data using these pipelines, too. And the `affine_registration` of `dipy` can be a benchmark, it should give 99.9% or something...

But I could have this totally backward, in which case the old code was probably wrong! But that is also how `morph.py` does things, so we must have some bugs on main, too...

Reply:
You were right about this, the affines needed to be combined, I'm not sure why the example worked previously though that's odd, looking into it.